File: data/scripts/pm_dynamic_vars.py
"""Pyodide dynamic-variable resolver — reads ``__pm_dynvar_json`` set by the host."""

from __future__ import annotations

import json
import random
import secrets
import string
import uuid
from datetime import UTC, datetime, timedelta
from typing import Any

_dynvar_data: dict[str, Any] | None = None


def _get_data() -> dict[str, Any]:
    global _dynvar_data
    if _dynvar_data is None:
        raw = globals().get("__pm_dynvar_json", "{}")
        if isinstance(raw, str):
            _dynvar_data = json.loads(raw)
        else:
            _dynvar_data = raw if isinstance(raw, dict) else {}
    return _dynvar_data


def _pick(pool: str, pools: dict[str, list[str]]) -> str:
    items = pools.get(pool, [])
    if not items:
        return ""
    return random.choice(items)


def _apply_rule(rule: dict[str, Any], pools: dict[str, list[str]]) -> str:
    kind = rule.get("rule", "")
    if kind == "uuid":
        return str(uuid.uuid4())
    if kind == "unixTime":
        return str(int(datetime.now(UTC).timestamp()))
    if kind == "isoTime":
        return datetime.now(UTC).isoformat().replace("+00:00", "Z")
    if kind == "intRange":
        lo = int(rule.get("min", 0))
        hi = int(rule.get("max", 1000))
        return str(secrets.randbelow(hi - lo + 1) + lo)
    if kind == "floatRange":
        lo = float(rule.get("min", 0))
        hi = float(rule.get("max", 1))
        dec = int(rule.get("decimals", 2))
        val = lo + (hi - lo) * random.random()
        return f"{val:.{dec}f}"
    if kind == "boolean":
        return secrets.choice(["true", "false"])
    if kind == "pick":
        return _pick(str(rule.get("pool", "")), pools)
    if kind == "picks":
        pool = str(rule.get("pool", ""))
        mn = int(rule.get("min", 1))
        mx = int(rule.get("max", mn))
        count = secrets.randbelow(mx - mn + 1) + mn
        return " ".join(_pick(pool, pools) for _ in range(count))
    if kind == "template":
        parts_out: list[str] = []
        for part in rule.get("parts", []):
            if isinstance(part, str):
                parts_out.append(part)
            elif isinstance(part, dict) and "pool" in part:
                parts_out.append(_pick(str(part["pool"]), pools))
        return "".join(parts_out)
    if kind == "hexColor":
        return f"#{secrets.token_hex(3)}"
    if kind == "ipv4":
        return ".".join(str(secrets.randbelow(256)) for _ in range(4))
    if kind == "ipv6":
        return ":".join(secrets.token_hex(2) for _ in range(8))
    if kind == "mac":
        return ":".join(f"{secrets.randbelow(256):02x}" for _ in range(6))
    if kind == "alphaNumeric":
        return secrets.choice(string.ascii_lowercase + string.digits)
    if kind == "password":
        alphabet = string.ascii_letters + string.digits + "!@#$%"
        return "".join(secrets.choice(alphabet) for _ in range(16))
    if kind == "semver":
        return f"{secrets.randbelow(10)}.{secrets.randbelow(20)}.{secrets.randbelow(100)}"
    if kind == "phone":
        return f"+1{secrets.randbelow(9000000000) + 1000000000:010d}"
    if kind == "phoneExt":
        return f"{_apply_rule({'rule': 'phone'}, pools)} x{secrets.randbelow(9000) + 1000}"
    if kind == "streetAddress":
        return f"{secrets.randbelow(9999) + 1} {_pick('streets', pools)}"
    if kind == "latitude":
        return f"{(random.random() * 180 - 90):.6f}"
    if kind == "longitude":
        return f"{(random.random() * 360 - 180):.6f}"
    if kind == "imageUrl":
        cat = str(rule.get("pool", "abstract"))
        return f"https://picsum.photos/seed/{secrets.token_hex(8)}/400/300?{cat}"
    if kind == "imageDataUri":
        return "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg=="
    if kind == "bankAccount":
        return "".join(str(secrets.randbelow(10)) for _ in range(10))
    if kind == "creditCardMask":
        return f"****-****-****-{secrets.randbelow(9000) + 1000:04d}"
    if kind == "bic":
        return "".join(secrets.choice(string.ascii_uppercase) for _ in range(8)) + "XX"
    if kind == "iban":
        return "GB" + "".join(str(secrets.randbelow(10)) for _ in range(20))
    if kind == "bitcoin":
        return "bc1" + secrets.token_hex(16)
    if kind == "companyName":
        return f"{_pick('lastNames', pools)} {_pick('companySuffixes', pools)}"
    if kind == "dateFuture":
        d = datetime.now(UTC) + timedelta(days=secrets.randbelow(365) + 1)
        return d.strftime("%Y-%m-%d")
    if kind == "datePast":
        d = datetime.now(UTC) - timedelta(days=secrets.randbelow(3650) + 1)
        return d.strftime("%Y-%m-%d")
    if kind == "dateRecent":
        d = datetime.now(UTC) - timedelta(days=secrets.randbelow(30))
        return d.strftime("%Y-%m-%d")
    if kind == "domainName":
        return f"{_pick('words', pools)}.{_pick('domains', pools)}"
    if kind == "exampleEmail":
        return f"user{secrets.randbelow(99999)}@example.com"
    if kind == "userName":
        return f"{_pick('firstNames', pools).lower()}{secrets.randbelow(999)}"
    if kind == "url":
        return f"https://{_pick('words', pools)}.{_pick('domains', pools)}"
    if kind == "fileName":
        return f"file_{secrets.token_hex(4)}.{_pick('fileExts', pools)}"
    if kind == "filePath":
        return f"/tmp/{_pick('words', pools)}.{_pick('fileExts', pools)}"
    if kind == "directoryPath":
        return f"/var/data/{_pick('words', pools)}"
    if kind == "price":
        return f"{secrets.randbelow(10000) / 100:.2f}"
    if kind == "ingVerb":
        v = _pick("hackerVerbs", pools)
        return v + "ing" if not v.endswith("ing") else v
    if kind == "loremSentence":
        return " ".join(_pick("loremWords", pools) for _ in range(secrets.randbelow(8) + 3)).capitalize() + "."
    if kind in ("loremSentences", "loremParagraph", "loremParagraphs", "loremText"):
        return _apply_rule({"rule": "loremSentence"}, pools)
    if kind == "loremSlug":
        return "-".join(_pick("loremWords", pools) for _ in range(3))
    if kind == "loremLines":
        return "\n".join(_apply_rule({"rule": "loremSentence"}, pools) for _ in range(3))
    if kind == "hackerAbbr":
        return "IO"
    if kind == "hackerPhrase":
        return (
            f"If we {_pick('hackerVerbs', pools)} the {_pick('hackerNouns', pools)}, "
            f"we can get to the {_pick('hackerNouns', pools)}."
        )
    return ""


def _pm_resolve_dynamic(name: str) -> str | None:
    """Resolve a Postman dynamic variable name for pyodide ``replace_in``."""
    key = str(name or "").strip()
    if not key:
        return None
    if not key.startswith("$"):
        key = f"${key}"
    data = _get_data()
    pools = data.get("pools", {})
    vars_map = data.get("vars", {})
    rule = vars_map.get(key)
    if not rule:
        return None
    return _apply_rule(rule, pools)

File: data/scripts/pm_dynamic_vars.js
// Postmark dynamic variable resolver — reads globalThis.__pm_dynvar (injected JSON).
(function () {
    function pick(pool, pools) {
        var items = pools[pool];
        if (!items || !items.length) return "";
        return items[Math.floor(Math.random() * items.length)];
    }

    function uuidV4() {
        var bytes = new Uint8Array(16);
        globalThis.crypto.getRandomValues(bytes);
        bytes[6] = (bytes[6] & 0x0f) | 0x40;
        bytes[8] = (bytes[8] & 0x3f) | 0x80;
        var hex = [];
        for (var i = 0; i < 16; i++) {
            hex.push(("0" + bytes[i].toString(16)).slice(-2));
        }
        return (
            hex.slice(0, 4).join("") +
            "-" +
            hex.slice(4, 6).join("") +
            "-" +
            hex.slice(6, 8).join("") +
            "-" +
            hex.slice(8, 10).join("") +
            "-" +
            hex.slice(10, 16).join("")
        );
    }

    function randInt(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    function applyRule(rule, pools) {
        var kind = rule.rule;
        if (kind === "uuid") return uuidV4();
        if (kind === "unixTime") return String(Math.floor(Date.now() / 1000));
        if (kind === "isoTime") return new Date().toISOString();
        if (kind === "intRange") return String(randInt(rule.min || 0, rule.max || 1000));
        if (kind === "floatRange") {
            var lo = rule.min || 0;
            var hi = rule.max || 1;
            var dec = rule.decimals || 2;
            var val = lo + (hi - lo) * Math.random();
            return val.toFixed(dec);
        }
        if (kind === "boolean") return Math.random() < 0.5 ? "true" : "false";
        if (kind === "pick") return pick(rule.pool, pools);
        if (kind === "picks") {
            var mn = rule.min || 1;
            var mx = rule.max || mn;
            var count = randInt(mn, mx);
            var out = [];
            for (var i = 0; i < count; i++) out.push(pick(rule.pool, pools));
            return out.join(" ");
        }
        if (kind === "template") {
            var parts = rule.parts || [];
            var s = "";
            for (var j = 0; j < parts.length; j++) {
                var p = parts[j];
                if (typeof p === "string") s += p;
                else if (p && p.pool) s += pick(p.pool, pools);
            }
            return s;
        }
        if (kind === "hexColor") return "#" + Math.floor(Math.random() * 0xffffff).toString(16).padStart(6, "0");
        if (kind === "ipv4") {
            return [0, 1, 2, 3].map(function () {
                return String(Math.floor(Math.random() * 256));
            }).join(".");
        }
        if (kind === "ipv6") {
            return [0, 1, 2, 3, 4, 5, 6, 7].map(function () {
                return Math.floor(Math.random() * 0xffff).toString(16);
            }).join(":");
        }
        if (kind === "mac") {
            return [0, 1, 2, 3, 4, 5].map(function () {
                return Math.floor(Math.random() * 256).toString(16).padStart(2, "0");
            }).join(":");
        }
        if (kind === "alphaNumeric") {
            var chars = "abcdefghijklmnopqrstuvwxyz0123456789";
            return chars.charAt(Math.floor(Math.random() * chars.length));
        }
        if (kind === "password") {
            var alpha = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%";
            var pw = "";
            for (var k = 0; k < 16; k++) pw += alpha.charAt(Math.floor(Math.random() * alpha.length));
            return pw;
        }
        if (kind === "semver") return randInt(0, 9) + "." + randInt(0, 19) + "." + randInt(0, 99);
        if (kind === "phone") return "+1" + String(randInt(1000000000, 9999999999));
        if (kind === "phoneExt") return applyRule({ rule: "phone" }, pools) + " x" + randInt(100, 999);
        if (kind === "streetAddress") return randInt(1, 9999) + " " + pick("streets", pools);
        if (kind === "latitude") return (Math.random() * 180 - 90).toFixed(6);
        if (kind === "longitude") return (Math.random() * 360 - 180).toFixed(6);
        if (kind === "imageUrl" || kind === "imageDataUri") {
            if (kind === "imageDataUri") {
                return "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==";
            }
            var cat = rule.pool || "abstract";
            var seed = Math.random().toString(36).slice(2, 10);
            return "https://picsum.photos/seed/" + seed + "/400/300?" + cat;
        }
        if (kind === "bankAccount") {
            var acct = "";
            for (var b = 0; b < 10; b++) acct += String(randInt(0, 9));
            return acct;
        }
        if (kind === "creditCardMask") return "****-****-****-" + String(randInt(1000, 9999));
        if (kind === "bic") {
            var bic = "";
            for (var c = 0; c < 8; c++) bic += String.fromCharCode(65 + randInt(0, 25));
            return bic + "XX";
        }
        if (kind === "iban") {
            var ib = "GB";
            for (var d = 0; d < 20; d++) ib += String(randInt(0, 9));
            return ib;
        }
        if (kind === "bitcoin") return "bc1" + Math.random().toString(16).slice(2, 18);
        if (kind === "companyName") return pick("lastNames", pools) + " " + pick("companySuffixes", pools);
        if (kind === "dateFuture" || kind === "datePast" || kind === "dateRecent") {
            var now = Date.now();
            var offset = randInt(1, kind === "dateRecent" ? 30 : 365);
            if (kind === "dateFuture") now += offset * 86400000;
            else now -= offset * 86400000;
            return new Date(now).toISOString().slice(0, 10);
        }
        if (kind === "domainName") return pick("words", pools) + "." + pick("domains", pools);
        if (kind === "exampleEmail") return "user" + randInt(0, 99999) + "@example.com";
        if (kind === "userName") return pick("firstNames", pools).toLowerCase() + randInt(0, 999);
        if (kind === "url") return "https://" + pick("words", pools) + "." + pick("domains", pools);
        if (kind === "fileName" || kind === "filePath" || kind === "directoryPath") {
            if (kind === "directoryPath") return "/var/data/" + pick("words", pools);
            if (kind === "filePath") return "/tmp/" + pick("words", pools) + "." + pick("fileExts", pools);
            return "file_" + Math.random().toString(36).slice(2, 6) + "." + pick("fileExts", pools);
        }
        if (kind === "price") return (randInt(0, 10000) / 100).toFixed(2);
        if (kind === "ingVerb") {
            var vb = pick("hackerVerbs", pools);
            return vb.endsWith("ing") ? vb : vb + "ing";
        }
        if (kind === "loremSentence") {
            var wds = [];
            for (var ls = 0; ls < randInt(3, 10); ls++) wds.push(pick("loremWords", pools));
            return wds.join(" ").replace(/^\w/, function (c) { return c.toUpperCase(); }) + ".";
        }
        if (kind === "loremSentences" || kind === "loremParagraph" || kind === "loremParagraphs" || kind === "loremText") {
            return applyRule({ rule: "loremSentence" }, pools);
        }
        if (kind === "loremSlug") {
            return [pick("loremWords", pools), pick("loremWords", pools), pick("loremWords", pools)].join("-");
        }
        if (kind === "loremLines") {
            return applyRule({ rule: "loremSentence" }, pools) + "\n" + applyRule({ rule: "loremSentence" }, pools);
        }
        if (kind === "hackerAbbr") return "IO";
        if (kind === "hackerPhrase") {
            return (
                "If we " +
                pick("hackerVerbs", pools) +
                " the " +
                pick("hackerNouns", pools) +
                ", we can get to the " +
                pick("hackerNouns", pools) +
                "."
            );
        }
        return "";
    }

    globalThis.__pm_resolveDynamic = function (name) {
        var key = String(name || "").trim();
        if (!key) return null;
        if (key.charAt(0) !== "$") key = "$" + key;
        var data = globalThis.__pm_dynvar;
        if (!data || !data.vars) return null;
        var rule = data.vars[key];
        if (!rule) return null;
        return applyRule(rule, data.pools || {});
    };
})();
